### ISSUE TITLE
What is this warning?

### ISSUE THREAD

**Issue by nagisa** [rust-lang/rust#32214]
```
src/librustc_mir/pretty.rs:20:36: 20:37 warning: whoops, no =?
src/librustc_mir/pretty.rs:20                                   w: &mut W) -> io::Result<()> {
                                                                 ^
```

**Comment by jonas-schievink** [comment]
https://github.com/rust-lang/rust/blame/master/src/libsyntax/parse/parser.rs#L4339

cc @nrc

**Comment by nagisa** [comment]
The code which caused this was similar to

```
pub fn write_mir_pretty<'a, W, I>(iter: &Iterator<Item=(&'a NodeId, &'a Mir<'a>), w: &mut W) -> io::Result<()>
```

note the missing `>` in iterator parameter list.

The warning with signature on one line:

```
src/librustc_mir/pretty.rs:19:84: 19:85 warning: whoops, no =?
src/librustc_mir/pretty.rs:19 pub fn write_mir_pretty<'a, W, I>(iter: &Iterator<Item=(&'a NodeId, &'a Mir<'a>), w: &mut W) -> io::Result<()>
```

**Comment by jonas-schievink** [comment]
This also causes the warning:
```rust
fn f<T:Iterator<Item=(), w>() {}
```
```
<anon>:1:27: 1:28 warning: whoops, no =?
<anon>:1 fn f<T:Iterator<Item=(), w>() {}
                                   ^
<anon>:1:27: 1:28 error: expected type, found `>`
<anon>:1 fn f<T:Iterator<Item=(), w>() {}
                                   ^
```

**Comment by durka** [comment]
https://github.com/rust-lang/rust/blob/master/src/libsyntax/parse/parser.rs#L4339

Seems like what it wants to convey is "expected associated type binding, found type parameter".

**Comment by nagisa** [comment]
@durka this is a parser and it has no/should not have notion of any such things.

I propose “error: expected `=`, found ...”, personally.

**Comment by nrc** [comment]
I'm a bit surprised that we bother to warn at all. Either it's an error and we should give an error (with a better message). Or we should not emit anything and rely on the error being caught later on (perhaps with a more informative error message).

In either case, we should explain *why* there is an error here - because you can't have unbound type variables after bound ones.
